client-sdk/ts-web/rt/src/callformat.ts

```diff
@@ -1,4 +1,146 @@
+import * as oasis from '@oasisprotocol/client';
+// @ts-expect-error missing declaration
+import * as deoxysii from 'deoxysii';
+import * as nacl from 'tweetnacl';
+// @ts-expect-error missing declaration
+import * as randomBytes from 'randombytes';
+
+import * as mraeDeoxysii from './mrae/deoxysii';
+import * as transaction from './transaction';
+import * as types from './types';
+
 /**
  * Call data key pair ID domain separation context base.
  */
 export const CALL_DATA_KEY_PAIR_ID_CONTEXT_BASE = 'oasis-runtime-sdk/private: tx';
+
+/**
+ *  EncodeConfig is call encoding configuration.
+ *  golang: oasis-sdk/client-sdk/go/callformat/callformat.go
+ *  rust:
+ */
+export interface EncodeConfig {
+    /**
+     * publicKey is an optional runtime's call data public key to use for encrypted call formats.
+     */
+    publicKey?: types.KeyManagerSignedPublicKey;
+}
+
+export interface MetaEncryptedX25519DeoxysII {
+    sk: Uint8Array;
+    pk: Uint8Array;
+}
+
+/**
+ * encodeCallWithNonceAndKeys encodes a call based on its configured call format.
+ * It returns the encoded call and any metadata needed to successfully decode the result.
+ */
+export function encodeCallWithNonceAndKeys(
```

Comment:
in some cases, we're making cryptography-related operations unnecessarily async, e.g. hashing a public key into an address. this is so that we could more easily change implementations in the future, if a more preferred library comes out exposing an async interface. or if the primitives we use later become available in the async web crypto interface and we could use that.

on the other hand, in other places, such as converting between binary and bech32 addresses, it's a synchronous interface. I think these places are the ones where it's a more niche kind of operation, and there's a single well established library for doing it, and that library does it synchronously.

I think this particular case is closer to the first scenario. but this is a late coming comment, and I understand we're interested in merging this right away to support some ongoing work elsewhere, so eh

Reply:
I am going to change the signature.